### ISSUE TITLE
Bash is required to run this release. Please ensure that it is installed.

### ISSUE THREAD

**Issue by ionull** [edib-tool/docker-edib-tool#15]
I created one simple project this month, however the built image cant be launched with this log: 

> Bash is required to run this release. Please ensure that it is installed.

BTW, the docker works fine on my previous project well :)

**Comment by ionull** [comment]
Seems that it was a distillery issue.
On 1.5.0, distillery requires one shell file in the path 'app/bin/your_app_loader.sh'

**Comment by DianaOlympos** [comment]
Let me have a look at this one, we should be able to add bash for the time being.

I know that @bitwalker is not a fan of depending on Bashism, so maybe it will solve itself on the long run.

**Comment by acgourley** [comment]
@ionull I am running into this too. Is there a workout, perhaps an older version of Distillery known to not require bash?

**Comment by acgourley** [comment]
Sorry - I see distillery 1.4 or below should work.

**Comment by DianaOlympos** [comment]
Yes distillery 1.4 will work.

I hope to solve that this weekend, but beware, it will probably grow the docker size quite a bit. We will see.

**Comment by acgourley** [comment]
I appreciate your work to make the size as small as possible - perhaps it would be enough to flag the problem so users can fix it more easily?

**Comment by bitwalker** [comment]
Adding bash to my `alpine-erlang` container added 1mb, so I think the bash requirement is a non issue. Unfortunately the older Distillery versions had an implicit requirement on bash, i.e., certain features, namely custom commands with arguments, would fail unexpectedly. If someone knows how to handle sourcing scripts with arguments, the bashism which pushes the bash requirement on us, I'm all ears, I'd prefer to stay compatible with POSIX sh if possible, but at this time I haven't found such a workaround.

**Comment by chvanikoff** [comment]
@bitwalker could you please update (including tags) `alpine-elixir` and `alpine-elixir-phoenix` as well?

**Comment by asaaki** [comment]
I've included `bash` in the artifact, maybe someone with the issue wants to try and verify/confirm:

~~`mix edib --edib edib/edib-tool:1.8.0-rc`~~
`mix edib --edib edib/edib-tool:1.8.0-rc2` (when [docker build](https://hub.docker.com/r/edib/edib-tool/builds/b8f6wbaol3zwbq2kmxmvukm/) is finished)

**Comment by acgourley** [comment]
On my project which was failing before with distillery 1.5.1, running `mix edib --edib edib/edib-tool:1.8.0-rc2` does create a working docker image! It is exactly 1 MB larger.

**Comment by asaaki** [comment]
@bitwalker So if I understand you correctly the reason to rely on bash is sourcing with arguments?

I tested under `docker run --rm -ti alpine:3.6 /bin/sh` the following:

```sh
/ # vi app.sh
```
```sh
#!/bin/sh
echo I am the main app and will source soureable.sh with arguments ...
. sourceable.sh some args via app.sh
echo This is supported by busybox sh.
```
```sh
/ # vi sourceable.sh
```
```sh
#!/bin/sh
echo "I am sourced by [ ${0##*/} ]"
echo "with [ $@ ] as parameter(s)"
```
```sh
/ # chmod +x app.sh
/ # ./app.sh
I am the main app and will source soureable.sh with arguments ...
I am sourced by [ app.sh ]
with [ some args via app.sh ] as parameter(s)
This is supported by busybox sh.
```

Inspired by <https://unix.stackexchange.com/a/284545/7949>.

Is this what you need and want?

-----

`Update` Scratch that, one of the other answers mentions the issue (I just missed it):

> You may be interested in the fact that the ability to source a script with arguments is a bashism.

— https://unix.stackexchange.com/a/5027/7949

**Comment by bitwalker** [comment]
@asaaki Yeah some implementations of `sh` support that, but as you noticed, it is nevertheless a bashism :). The POSIX `sh` spec does not allow for it, but as far as I'm aware, most systems alias `sh` to `bash`, and others, like busybox `sh`, support it, but the big exception is `dash`, the `sh` implementation in Debian/Ubuntu. If we want to be truly portable, we have to support the least common denominator. 

That said, the problem solved by sourcing scripts with arguments in Distillery is that the functions and environment defined in the sourcing script are available to the sourced script. This allows custom commands/hooks/etc., to use the helpers Distillery defines without having to source the right scripts manually (and I don't like that anyway, since it makes those scripts part of the public API, and I would prefer that to be transparent to users so I'm free to make changes internally). As far as I'm aware (and in the testing I did, which was by no means comprehensive, but was enough to indicate there were problems), neither `exec` nor invoking scripts with arguments (e.g., `/path/to/some/script.sh arg1 arg2` vs `. /path/to/some/script.sh arg1 arg2`) carries function definitions into the new environment. If we can find a way other than sourcing to accomplish this, which is POSIX-compatible, then we can remove the dependency on `bash`. I have hope that it is possible and that I missed something when I was researching the problem, but as I did not have time to dig deep and then subsequently test on all the platforms in common use by users of Distillery, I decided it made more sense to explicitly require `bash` for now, to ensure that what is already there is fully supported, and then wait for someone with more shell knowledge than me to point out what I missed, or at least an alternative that solves the same problem, or once I find time, perhaps solve it myself. Since bash is mostly ubiquitous anyway, outside of a few platforms (Solaris and Alpine are the only ones that I believe are in use by users of Distillery, and I could be mistaken about Solaris), I felt it wasn't too pressing of an issue.
